sdks/java/io/redis/src/main/java/org/apache/beam/sdk/io/redis/RedisIO.java

```diff
@@ -0,0 +1,248 @@
+
+    @Override
+    public void validate(PipelineOptions pipelineOptions) {
+      checkState(connectionConfiguration() != null || redisService() != null,
+          "RedisIO.read() requires a connectionConfiguration to be set "
+              + "withConnection(connectionConfiguration) or a service to be set withRedisService"
+              + "(service)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      if (connectionConfiguration() != null) {
+        connectionConfiguration().populateDisplayData(builder);
+      }
+    }
+
+    @Override
+    public PCollection<KV<String, String>> expand(PBegin input) {
+      return input.getPipeline()
+          .apply(org.apache.beam.sdk.io.Read.from(
+              new RedisSource(
+                  keyPattern(),
+                  new SerializableFunction<PipelineOptions, RedisService>() {
+                    @Override
+                    public RedisService apply(PipelineOptions pipelineOptions) {
+                      return getRedisService(pipelineOptions);
+                    }
+                  }, null)));
+    }
+
+    /**
+     * Helper function to either get a fake/mock Redis service provided by
+     * {@link #withRedisService(RedisService)} or creates and returns an implementation of a
+     * concrete Redis service dealing with an actual Redis server.
+     */
+    @VisibleForTesting
+    RedisService getRedisService(PipelineOptions pipelineOptions) {
+      if (redisService() != null) {
+        return redisService();
+      }
+      return new RedisServiceImpl(connectionConfiguration());
+    }
+
+  }
+
+  /**
+   * A bounded source reading key-value pairs from a Redis server.
+   */
+  @VisibleForTesting
+  protected static class RedisSource extends BoundedSource<KV<String, String>> {
+
+    protected final String keyPattern;
+    private final SerializableFunction<PipelineOptions, RedisService> serviceFactory;
+    protected final RedisService.RedisNode node;
+
+    protected RedisSource(String keyPattern,
+                          SerializableFunction<PipelineOptions, RedisService> serviceFactory,
+                          RedisService.RedisNode node) {
+      this.keyPattern = keyPattern;
+      this.serviceFactory = serviceFactory;
+      this.node = node;
+    }
+
+    @Override
+    public List<RedisSource> split(long desiredBundleSizeBytes,
+                                              PipelineOptions pipelineOptions) throws IOException {
+      if (serviceFactory.apply(pipelineOptions).isClusterEnabled()) {
+        LOG.info("Cluster detected");
+        List<RedisSource> redisSources = new ArrayList<>();
+
+        List<RedisService.RedisNode> nodes = serviceFactory.apply(pipelineOptions)
+            .getClusterNodes();
+
+        int slot = serviceFactory.apply(pipelineOptions).getKeySlot(keyPattern);
+
+        for (RedisService.RedisNode node : nodes) {
+          if (node.startSlot < slot && node.endSlot > slot) {
```

Comment:
Is this correct? Shouldn't it be something like `node.startSlot <= slot && slot < node.endSlot`?